nntrainer/utils/node_exporter.h

```diff
@@ -72,16 +100,51 @@ class Exporter {
    * @tparam T appropriate return type regarding the export method
    * @return T T
    */
-  template <ExportMethods methods, typename T = return_type<methods>>
-  T get_result() {
-    if (!is_exported) {
-      throw std::invalid_argument("This exporter is not exported anything yet");
-    }
-    /** NYI!! */
-  }
+  template <ExportMethods methods,
+            typename T = typename return_type<methods>::type>
+  const T &get_result();
 
 private:
-  bool is_exported;
+  /**
+   * @brief base case of iterate_prop, iterate_prop iterates the given tuple
+   *
+   * @tparam I size of tuple(automated)
+   * @tparam Callable generic lambda to be called during iteration
+   * @tparam Ts types from tuple
+   * @param c callable gerneric labmda
+   * @param tup tuple to be iterated
+   * @return void
+   */
+  template <size_t I = 0, typename Callable, typename... Ts>
+  typename std::enable_if<I == sizeof...(Ts), void>::type
+  iterate_prop(Callable &&c, const std::tuple<Ts...> &tup) {
+    // end of recursion;
+  }
+
+  /**
+   * @brief base case of iterate_prop, iterate_prop iterates the given tuple
+   *
+   * @tparam I size of tuple(automated)
+   * @tparam Callable generic lambda to be called during iteration
+   * @tparam Ts types from tuple
+   * @param c callable gerneric labmda
+   * @param tup tuple to be iterated
+   * @return not used
+   */
+  template <size_t I = 0, typename Callable, typename... Ts>
+  typename std::enable_if<(I < sizeof...(Ts)), void>::type
+  iterate_prop(Callable &&c, const std::tuple<Ts...> &tup) {
+    c(std::get<I>(tup), I);
+
+    iterate_prop<I + 1>(c, tup);
+  }
+
+  std::vector<std::pair<std::string, std::string>>
+    stored_result; /**< stored result */
```

Comment:
This is fine for now, but later you can consider if this is needed. This is dangerous as there is no way to know if this stored_result is valid now, as some properties might have changed since then.

Reply:
I think it is sensible that returned value of get_property can always be invalidated after set_property is called.

I think it is also sensible not to call get_property in the middle of set_property.
But If we need to make it sure, we can have a mutex I guess.

Reply:
The scenario you described is a legitimate one. However, there can be another scenario too.
```
model.train('batch_size=32');
model.saveResult();
model.train('batch_size=16');
model.getResult(); // <- batch size in this result will be wrong and not sure, if mutex can resolve this
```